File: 0001-0050/0017/solution.py
from typing import List


class Solution:
    ans = []
    strs = [
        "", "", "abc", "def",
        "ghi", "jkl", "mno",
        "pqrs", "tuv", "wxyz"
    ]

    def letterCombinations(self, digits: str) -> List[str]:
        self.ans = []
        if not digits:
            return self.ans
        self.dfs(digits, 0, "")
        return self.ans

    def dfs(self, digits, idx, combine):
        if idx == len(digits):
            self.ans.append(combine)
            return
        s = self.strs[int(digits[idx])]
        for i in range(len(s)):
            combine += s[i]
            self.dfs(digits, idx + 1, combine)
            combine = combine[:-1]


class Example:
    def __init__(self, digits, ans):
        self.digits = digits
        self.ans = ans


def main():
    examples = [
        Example("23", ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]),
        Example("", []),
        Example("2", ["a", "b", "c"])
    ]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.letterCombinations(example.digits)
        if ans == example.ans:
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()

File: 0001-0050/0017/solution.go
package main

import (
    "fmt"
    "reflect"
)

var ans []string
var strs = []string{
    "", "", "abc", "def",
    "ghi", "jkl", "mno",
    "pqrs", "tuv", "wxyz",
}

func letterCombinations(digits string) []string {
    ans = []string{}
    if len(digits) == 0 {
        return ans
    }
    dfs(digits, 0, "")
    return ans
}

func dfs(digits string, idx int, combine string) {
    if idx == len(digits) {
        ans = append(ans, combine)
        return
    }
    s := strs[digits[idx]-'0']
    for i := 0; i < len(s); i++ {
        combine += string(s[i])
        dfs(digits, idx+1, combine)
        combine = combine[:len(combine)-1]
    }
}

type Example struct {
    digits string
    ans    []string
}

func main() {
    examples := []Example{
        {"23", []string{"ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"}},
        {"", []string{}},
        {"2", []string{"a", "b", "c"}},
    }
    for i, example := range examples {
        ans := letterCombinations(example.digits)
        if reflect.DeepEqual(ans, example.ans) {
            fmt.Printf("PASS: CASE %d\n", i)
        } else {
            fmt.Printf("FAIL: CASE %d\n", i)
        }
    }
}
